### PR TITLE
Style package with new {mlr.styler} package

### PR DESCRIPTION
**Pull request by pat-s**
This is a first diff of the new {[mlr.styler](https://github.com/mlr-org/styler.mlr)} package (still private) by @lorenzwalthert.
Lorenz put a [lot of time](https://github.com/mlr-org/styler.mlr/pull/1) into our mlr-style, not only by creating this new package but also in the past.

It should make updating + maintenance more easy and should also serve as an exemplary package for additional styles next to the `tidyverse_style`.
Beforehand we used a [fork of mine](https://github.com/pat-s/styler/tree/mlr-style).

@mllg I am not sure if you use {styler} in practice for {mlr3} but your input on specific re-occurring patterns which are changed in the diff is appreciated.
Remember that we can create the style guide to be passive/non-invasive so that in some cases multiple solutions are allowed / not force styled.

Some years have passed since we evaluated the initial guide. If some of your coding patterns have changed since there, it would now be a good time to discuss these here.

Also cc @berndbischl @mb706 @be-marc @pfistfl @jakob-r @RaphaelS1 (your input is appreciated/desired as well of course as the style applies to all mlr-org repos. {mlr3} is just the demo-repo for now.).

@lorenzwalthert
If there are any new transformers that you know of which would be possibly helpful here, just suggest them :)

### CODE REVIEW

File: R/BenchmarkResult.R
Comment (pat-s) [inline R/BenchmarkResult.R:369]: @mllg 

IIRC we discussed about this pattern/change the last time already and agreed to allow both options?

File: R/DataBackend.R
Comment (pat-s) [inline R/DataBackend.R:30]: @mllg Is this something we want to have?

File: R/LearnerClassifDebug.R
Comment (pat-s) [inline R/LearnerClassifDebug.R:65]: @lorenzwalthert 

IIRC we had support for this in the old fork - was there a specific transformer or did we do a hacky island solution here?

File: R/BenchmarkResult.R
Comment (pat-s) [reply R/BenchmarkResult.R:369]: Tested this on another repo and in our old mlr-style this was not changed by {styler} -> @lorenzwalthert can we remove this styling? :)

**Comment by lorenzwalthert** [comment]
@pat-s thanks, will look into it tomorrow. Alternatively, you could make one PR per scope, starting with `spaces`, instead of all of it together. Maybe it’s easier to review.

File: R/LearnerClassifDebug.R
Comment (lorenzwalthert) [reply R/LearnerClassifDebug.R:65]: alignment is supported in general with {styler}, but only with right-alignment:

``` r
styler.mlr::style_text(
  'call(
  x  =   2, 
  uz = 333
  )
  '
)
#> call(
#>   x  =   2,
#>   uz = 333
#> )
```

<sup>Created on 2021-04-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

File: R/BenchmarkResult.R
Comment (lorenzwalthert) [reply R/BenchmarkResult.R:369]: Indeed, there was a transformer for that,but for some reason, it was commented out in the revision of your fork I used when c/p the transformers. Adressed in https://github.com/mlr-org/styler.mlr/pull/7.

File: R/DataBackend.R
Comment (lorenzwalthert) [reply R/DataBackend.R:30]: You did not have this previously. Fix suggested in https://github.com/mlr-org/styler.mlr/pull/9/files.

File: R/ResamplingBootstrap.R
Comment (lorenzwalthert) [inline R/ResamplingBootstrap.R:73]: I note there are a lot of these line breaks removed with `strict = FALSE`. This is due to `styler:::remove_line_break_in_fun_call()`, which removes empty lines in function calls, but only if no comment is around, as discussed in https://github.com/r-lib/styler/issues/632 and https://github.com/r-lib/styler/issues/627. We can either remove this rule or set a minimal number of lines that are required in order to assume this is a *big call* and the line breaks are used to *group* different arguments (similar to empty lines we insert at the top of a function declaration if the body of the function has more than k lines) and keep the blank lines for these situations.

Comment (mllg) [reply R/ResamplingBootstrap.R:73]: Extra newlines normally serve a purpose, I would also opt to disable this.

File: R/BenchmarkResult.R
Comment (mllg) [inline R/BenchmarkResult.R:221]: I'm not a big fan of enforcing a newline here. Any reason to keep this?

Comment (lorenzwalthert) [reply R/BenchmarkResult.R:221]: I can't find a reference anymore in the style guide but at some point the idea was to leave a blank line when the function declaration is long. Assuming you have a function declaration somewhere that has a blank line at the top, would you want to remove it or simply not touch it? If we don't touch it, we also can't touch 
```r
{ 
  # code 
}
``` 
easily (because of the inner workings of styler). A compromise would be to allow up either one or zero blank lines in any of these two curly brace expressions, but not more.

File: R/LearnerClassifDebug.R
Comment (lorenzwalthert) [reply R/LearnerClassifDebug.R:65]: I suggest to manually align them to one of the supported alignments in https://styler.r-lib.org/articles/detect-alignment.html to make sure styler does not *destroy* the formatting of deliberately aligned code.

**Comment by RaphaelS1** [comment]
Just a short comment for now. Points 11 and 17 in the mlr style guide seem to have evolved to the opposite so `return` statements are rare and we decided to explicitly use `::` for most packages except mlr3 and paradox (e.g. we tend to explicitly write `mlr3misc::`)

File: R/LearnerClassifDebug.R
Comment (lorenzwalthert) [reply R/LearnerClassifDebug.R:65]: Left alignment support on the way: https://github.com/r-lib/styler/pull/774

File: R/ResamplingBootstrap.R
Comment (lorenzwalthert) [reply R/ResamplingBootstrap.R:73]: Addressed in https://github.com/mlr-org/styler.mlr/issues/10.

File: R/BenchmarkResult.R
Comment (pat-s) [reply R/BenchmarkResult.R:369]: Is this a possible misunderstanding?

I meant to say that we want to allow the 

```
    }
  )
``` 

pattern. And avoid a styling into `})`.
I think this is why I commented out this transformer at some point.

I see that in the pull you explicitly commented it in again.
Assuming that it was commented out in this current draft - why did this styling happen then? 🤔 
Or am I on the wrong path?

File: R/DataBackend.R
Comment (pat-s) [reply R/DataBackend.R:30]: Thanks 🎉

File: R/LearnerClassifDebug.R
Comment (pat-s) [reply R/LearnerClassifDebug.R:65]: great 🎉

File: R/ResamplingBootstrap.R
Comment (pat-s) [reply R/ResamplingBootstrap.R:73]: 🎉

File: R/BenchmarkResult.R
Comment (pat-s) [reply R/BenchmarkResult.R:221]: > I can't find a reference anymore in the style guide

The style guide might also not be 100% up-to-date 😅 

I think being deliberate here is probably best - i.e. allowing both.

> A compromise would be to allow up either one or zero blank lines in any of these two curly brace expressions, but not more.

Sounds like a good compromise to me (if that is easier for {styler} also). @mllg?

I think we never want to have two blank lines anywhere and any of zero or one is fine?

**Comment by pat-s** [comment]
> Just a short comment for now. Points 11 and 17 in the mlr style guide seem to have evolved to the opposite so return statements are rare and we decided to explicitly use :: for most packages except mlr3 and paradox (e.g. we tend to explicitly write mlr3misc::)

Yeah, the guide is probably not up-to-date or followed fully in practice.
I think {styler} does not touch semantic things like adding/removing `return()` statements IIRC.

The same goes for explicit namespacing I think - this is probably also not trivial and would require quite some magic to infer this from the function name.

But yeah, we should probably adjust this in the guide.

File: R/BenchmarkResult.R
Comment (lorenzwalthert) [reply R/BenchmarkResult.R:369]: Yup, there was a misunderstanding (and also merge conflicts). We want to activate the rule that adds a line break if there is none and remove the rule that removes the line break. Force pushed to https://github.com/mlr-org/styler.mlr after rebasing on main.

File: R/LearnerClassifDebug.R
Comment (lorenzwalthert) [reply R/LearnerClassifDebug.R:64]: You need to have the devel version of styler with `9f9bdff` at least yes. I am adding a minimal version requirement to {styler.mlr} aud bump the devel version in {styler}.

Comment (lorenzwalthert) [reply R/LearnerClassifDebug.R:64]: Done. I suggest to re-style.

File: R/TaskClassif_penguins.R
Comment (pat-s) [inline R/TaskClassif_penguins.R:38]: @lorenzwalthert DO you have an idea why this alignment is not honoured?

File: R/bibentries.R
Comment (pat-s) [inline R/bibentries.R:26]: @lorenzwalthert DO you have an idea why this alignment is not honoured?

File: R/mlr_reflections.R
Comment (pat-s) [inline R/mlr_reflections.R:79]: @lorenzwalthert DO you have an idea why this alignment is not honoured?

File: R/LearnerClassifDebug.R
Comment (pat-s) [reply R/LearnerClassifDebug.R]: 👍 

resetted and restyled - some more alignment takes it seems :)

File: R/TaskClassif_penguins.R
Comment (lorenzwalthert) [reply R/TaskClassif_penguins.R:38]: Because spacing is only honoured one level deep. This is two level deep (function call in function call). Sorry, styler can't do this. Use `strict = FALSE` if you want to preserve one more more space here (but it comes with other implications as shown in the dedicated vignette).

File: R/bibentries.R
Comment (lorenzwalthert) [reply R/bibentries.R:26]: Because title is not aligned xD. You must add some spaces after `title` before `=`.

File: R/mlr_reflections.R
Comment (lorenzwalthert) [reply R/mlr_reflections.R:79]: Because left alignment only works with `=`, not with `~`. See https://github.com/lorenzwalthert/mlr3/pull/2#discussion_r619304282

**Review by lorenzwalthert — CHANGES_REQUESTED** [review]

File: R/bibentries.R
Comment (pat-s) [reply R/bibentries.R:26]: Ah yeah! I only looked at the green part of the diff I guess.

File: R/mlr_reflections.R
Comment (lorenzwalthert) [reply R/mlr_reflections.R:79]: I tried for a few hours and I did not get to work it easily... Maybe I can look into it again since we are on it.

Comment (pat-s) [reply R/mlr_reflections.R:79]: This is really niche so really think about if you want to do this :)

Would be great and appreciated of course though 👍

Comment (lorenzwalthert) [reply R/mlr_reflections.R:79]: Ok, I found a way to do it and merged it into master: https://github.com/r-lib/styler/pull/785. After the other alignment things were implemented properly, it seemed much easier than I thought initially. Are you happy now? xD

Comment (pat-s) [reply R/mlr_reflections.R:79]: Thanks! Just restyled but it seems it not yet working 🙈

Comment (lorenzwalthert) [reply R/mlr_reflections.R:79]: Because you have an extra wurst with `.key = '...'` -.-. Fix on the way...

Comment (lorenzwalthert) [reply R/mlr_reflections.R:79]: Please try again xD

**Comment by pat-s** [comment]
@lorenzwalthert The latest diff looks good to me! Thanks a lot for your help and persistence here 🙏 

@mllg Are you fine with the diff and hence the style guide? If so, we can merge and I would update the documentation and inform all devs.

File: R/mlr_reflections.R
Comment (lorenzwalthert) [reply R/mlr_reflections.R]: This seems resolved with latest styling.

**Comment by lorenzwalthert** [comment]
What’s failing here @pat-s ?

**Comment by pat-s** [comment]
Some git checkout issues, seems random. I think we can ignore it.

Styling is fine 👍

**Comment by lorenzwalthert** [comment]
I think it was a GitHub outage. And now we have a merge conflict. I suggest to resolve it, retrigger styling and then get this merged before more merge conflicts accumulate.

**Comment by pat-s** [comment]
I'd like to have (at least) an approval by @mllg

**Comment by mllg** [comment]
Looks good to me. Thanks @pat-s and @lorenzwalthert for all the work you've put into this.

**Pull request merged**
